### DISCUSSION TITLE
How to use atomWithStorage with suspense in web with localStorage + reducer?

### DISCUSSION THREAD

**Discussion by jetaggart** [pmndrs/jotai#1837] [Q&A]
Hello, I'm struggling to understand how I can get a suspense fallback to fire when waiting for an atom to be loaded from localStorage. I have this set up:

```typescript
const baseAtom = atomWithStorage<CartState>("cart", {_type: "CartState", items: []})
const cartAtom = atom(
  (get) => get(baseAtom),
  async (get, set, action: CartAction) => {
    set(baseAtom, await reducer(get(baseAtom), action))
  }
)
export const useCart = () => useAtom(cartAtom)
```

I'd like the default value to only be returned if nothing is found in storage and a suspense fallback to fire if it's waiting for the baseAtom to load. Is this possible?

I'm trying to have a "Loading" state when the cart is first initialized on the page. Right now it flickers no items until it loads from storage.

**Comment by dai-shi** [comment]
To have a suspense loading, you need to store a promise. It's possible if the base atom is a normal atom.
```ts
const baseAtom = atom<CartState | Promise<CartState>>({_type: "CartState", items: []})
const cartAtom = atom(
  (get) => get(baseAtom),
  (get, set, action: CartAction) => {
    set(baseAtom, Promise.resolve(get(baseAtom)).then((prev) => reducer(prev, action)))
  }
)
export const useCart = () => useAtom(cartAtom)
```

Now, it's a bit tricky if you want to do it with storage because you can't store a promise. There are several workarounds: a) make custom storage to handle promises, and b) create two base atoms for storage and promise.

**Reply by jetaggart** [reply]
Thank you for the response. `atomWithStorage` loads the default value first then later loads from storage, so there's an async action happening internal to the atom, correct? Is there any way to hook into that state either by suspense boundary or a loading state? My intuition says that if `atomWithStorage` is using a promise internally, and it is in a non-resolved state until the storage returns, and if it finds nothing it uses the default value, otherwise it uses what's in storage. I didn't expect it to use the default value then update with what is found in storage. Do I have my mental model wrong?

**Reply by dai-shi** [reply]
I'm not sure if I get your point. Is everything okay for you without `atomWithStorage`?

**Reply by jetaggart** [reply]
If you look at this example from the docs: https://codesandbox.io/s/jotai-persistence-vuwi7, you can see a flicker if you change to dark mode. If you log the values it goes `light` then `dark` with two values being reported, the default then the value from storage:

<img width="305" alt="Screenshot 2023-03-16 at 10 14 20 AM" src="https://user-images.githubusercontent.com/161737/225683778-0dd23274-8ad5-4d22-bd1e-ba9904fd00cd.png">

 I expect `atomWithStorage` to only use the default value if nothing is found in storage, and some sort of loading indication when it's waiting to read from storage instead of the default value. I'd want to use a suspense boundary to fallback when the atom is loading from storage. It appears the atom loads the default first, then later loads from storage, with no way to hook into that lifecycle.

FWIW, I ended up with code like this, but it doesn't feel right:

```typescript
const baseAtom = atomWithStorage<CartState | "loading">("cart", "loading")
const cartAtom = atom(
  (get) => get(baseAtom) || {_type: "CartState", items: []},
  async (get, set, action: CartAction) => {
    set(baseAtom, await reducer(get(baseAtom), action))
  }
)
```

**Reply by dai-shi** [reply]
Okay, `atomWithStorage` does flicker. That's how it's designed in v2.
If you need what you want, you would need a different util.
`atomWithLocalStorage` would be something simple shown in the persistence recipe.
`atomWithAsyncStorage` would be something going on in #1784.

**Reply by jetaggart** [reply]
Ah yeah that's exactly it. Thank you!
